src/OrleansRuntime/Streams/QueueBalancer/LeaseBasedQueueBalancer.cs

```diff
@@ -0,0 +1,345 @@
+        private int minimumResponsibilty;
+        private int maximumRespobsibility;
+        private Logger logger;
+
+        public LeaseBasedQueueBalancer(ILeaseProvider leaseProvider, ISiloStatusOracle siloStatusOracle, IDeploymentConfiguration deploymentConfig, Factory<string, Logger> loggerFac)
+            : base()
+        {
+            this.leaseProvider = leaseProvider;
+            this.deploymentConfig = deploymentConfig;
+            this.siloStatusOracle = siloStatusOracle;
+            this.myQueues = new List<AcquiredQueue>();
+            this.immatureSilos = new ConcurrentDictionary<SiloAddress, bool>();
+            this.isStarting = true;
+            this.logger = loggerFac(this.GetType().Name);
+            // register for notification of changes to silo status for any silo in the cluster
+            this.siloStatusOracle.SubscribeToSiloStatusEvents(this);
+            // record all already active silos as already mature. 
+            // Even if they are not yet, they will be mature by the time I mature myself (after I become !isStarting).
+            foreach (var silo in siloStatusOracle.GetApproximateSiloStatuses(true).Keys.Where(s => !s.Equals(siloStatusOracle.SiloAddress)))
+            {
+                immatureSilos[silo] = false;     // record as mature
+            }
+        }
+
+        public override Task Initialize(string strProviderName, IStreamQueueMapper queueMapper, TimeSpan siloMaturityPeriod)
+        {
+            if (queueMapper == null)
+            {
+                throw new ArgumentNullException("queueMapper");
+            }
+            this.allQueues = new ReadOnlyCollection<QueueId>(queueMapper.GetAllQueues().ToList());
+            this.siloMaturityPeriod = siloMaturityPeriod;
+            NotifyAfterStart().Ignore();
+            //make lease renew frequency to be every half of lease time, to avoid renew failing due to timing issues, race condition or clock difference. 
+            this.renewLeaseTimer = new AsyncTaskSafeTimer(this.MaintainAndBalanceQueues, null, this.siloMaturityPeriod, this.leaseLength.Divide(2));
+            //try to acquire maximum leases every leaseLength 
+            this.tryAcquireMaximumLeaseTimer = new AsyncTaskSafeTimer(this.AcquireLeaseToMeetMaxResponsibilty, null, this.siloMaturityPeriod, this.leaseLength);
+            //Selector default to round robin selector now, but we can make a further change to make selector configurable if needed.  Selector algorithm could 
+            //be affecting queue balancing stablization time in cluster initializing and auto-scaling
+            this.queueSelector = new RoundRobinSelector<QueueId>(this.allQueues.ToList());
+            return MaintainAndBalanceQueues(null);
+        }
+
+        public override IEnumerable<QueueId> GetMyQueues()
+        {
+            return this.myQueues.Select(queue => queue.QueueId);
+        }
+
+        private async Task MaintainAndBalanceQueues(object state)
```

Comment:
I think we need to call NotifyListeners at the end of these timer calls any time myQueues changes.